### DISCUSSION TITLE
for Queue::write_texture, does `data_layout` refer to the source image, or the destination image?

### DISCUSSION THREAD

**Discussion by spencerkohan** [gfx-rs/wgpu#3172] [Q&A] [answered]
I'm working on packing a texture atlas.  So I have image data for my source images, as `Vec<u8>` and I want to write this into a 2D texture array.

I see I can do this with the ['Queue::write_texture' method](https://wgpu.rs/doc/wgpu/struct.Queue.html#method.write_texture).

This method takes an [ImageDataLayout](https://wgpu.rs/doc/wgpu/struct.ImageDataLayout.html#) as an argument, which allows me to specify the bytes per row, and rows per image of *something*, but it's not clear to me whether this is the source image (bytes of the bitmap being copied) or the destination image (image the bitmap bytes are being copied to).

So for instance, if I have a 2x2 image:

```
let bitmap: &[u8] = &[0, 1, 1, 0];
```

And I want to copy this into the top left corner of a 4x4 image, how would I achieve this?

edit:

for instance in [Vulkan](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VkBufferImageCopy.html) you would specify the image subresource range you want to copy to

**Comment by cwfitzgerald** [comment, marked as answer]
It refers to the layout in the buffer. For example to copy that you'd have a bytes per row of 2, an extent of 2x2x1 an an origin (in the image copy struct) of 0,0,0

We'd be open to a PR that improves the verbiage here.

**Reply by kpreid** [reply]
I created PR #3173 to expand the documentation.
